pkg/services/ngalert/accesscontrol/silences.go

```diff
@@ -0,0 +1,252 @@
+	// asserts that user has access to create silences at all
+	createSomeRuleSilenceEvaluator = ac.EvalAll(
+		readSomeSilenceEvaluator,
+		ac.EvalAny(
+			ac.EvalPermission(instancesCreate),
+			ac.EvalPermission(silenceCreate)),
+	)
+	// asserts that user has access to create silences in a specific folder
+	createRuleSilenceEvaluator = func(uid string) ac.Evaluator {
+		return ac.EvalAll(
+			ac.EvalAny(
+				ac.EvalPermission(instancesCreate),
+				ac.EvalPermission(silenceCreate, dashboards.ScopeFoldersProvider.GetResourceScopeUID(uid)),
+			),
+			readRuleSilenceEvaluator(uid),
+		)
+	}
+
+	// shortcut assertion to check if user can update any silence
+	updateAnySilenceEvaluator = ac.EvalAll(ac.EvalPermission(instancesWrite), readAllSilencesEvaluator)
+	// asserts that user has access to update global silences
+	updateGlobalSilenceEvaluator = ac.EvalAll(ac.EvalPermission(instancesWrite), readSomeSilenceEvaluator)
+	// asserts that user has access to update silences at all
+	updateSomeRuleSilenceEvaluator = ac.EvalAll(
+		readSomeSilenceEvaluator,
+		ac.EvalAny(
+			ac.EvalPermission(instancesWrite),
+			ac.EvalPermission(silenceWrite)),
+	)
+	// asserts that user has access to create silences in a specific folder
+	updateRuleSilenceEvaluator = func(uid string) ac.Evaluator {
+		return ac.EvalAll(
+			ac.EvalAny(
+				ac.EvalPermission(instancesWrite),
+				ac.EvalPermission(silenceWrite, dashboards.ScopeFoldersProvider.GetResourceScopeUID(uid)),
+			),
+			readRuleSilenceEvaluator(uid),
+		)
+	}
+)
+
+type Silence interface {
+	GetRuleUID() *string
+}
+
+type RuleUIDToNamespaceStore interface {
+	GetNamespacesByRuleUID(ctx context.Context, orgID int64, uids ...string) (map[string]string, error)
+}
+
+type SilenceService struct {
+	genericService
+	store RuleUIDToNamespaceStore
+}
+
+func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *SilenceService {
+	return &SilenceService{
+		genericService: genericService{
+			ac: ac,
+		},
+		store: store,
+	}
+}
+
+// FilterByAccess filters the given list of silences based on the access control permissions of the user.
+// Global silence (one that is not attached to a particular rule) is considered available to all users.
+// For silences that are not attached to a rule, are checked against authorization.
+// This method is more preferred when many silences need to be checked.
+func (s SilenceService) FilterByAccess(ctx context.Context, user identity.Requester, silences ...Silence) ([]Silence, error) {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	if err != nil || canAll { // return early if user can either read all silences or there is an error
+		return silences, err
+	}
+	canSome, err := s.HasAccess(ctx, user, readSomeSilenceEvaluator)
+	if err != nil || !canSome {
+		return nil, err
+	}
+	result := make([]Silence, 0, len(silences))
+	keys := make(map[string][]Silence, len(silences))
```

Comment:
NIT: Can we improve the `keys` variables here to be more descriptive. Maybe `ruleSilences`, for example.